sanitizers/src/main/java/com/code_intelligence/jazzer/sanitizers/ScriptEngineInjection.java

```diff
@@ -0,0 +1,146 @@
+  @MethodHook(type = HookType.REPLACE, targetClassName = "javax.script.ScriptEngine",
+      targetMethod = "eval",
+      targetMethodDescriptor = "(Ljava/io/Reader;Ljavax/script/ScriptContext;)Ljava/lang/Object;")
+  public static Object
+  checkScriptEngineExecuteScriptContext(
+      MethodHandle method, Object thisObject, Object[] arguments, int hookId) throws Throwable {
+    if (evalWithScriptContext == null) {
+      evalWithScriptContext =
+          methodFromObject(thisObject, "eval", "java.lang.String", "javax.script.ScriptContext");
+    }
+    if (evalWithScriptContext.isPresent()) {
+      String content = checkScriptContent(arguments[0], hookId);
+      return evalWithScriptContext.get().invoke(thisObject, content, arguments[1]);
+    } else {
+      return method.invokeWithArguments(thisObject, arguments);
+    }
+  }
+
+  @MethodHook(type = HookType.REPLACE, targetClassName = "javax.script.ScriptEngine",
+      targetMethod = "eval",
+      targetMethodDescriptor = "(Ljava/lang/String;Ljavax/script/Bindings;)Ljava/lang/Object;")
+  @MethodHook(type = HookType.REPLACE, targetClassName = "javax.script.ScriptEngine",
+      targetMethod = "eval",
+      targetMethodDescriptor = "(Ljava/io/Reader;Ljavax/script/Bindings;)Ljava/lang/Object;")
+  public static Object
+  checkScriptEngineExecuteBindings(
+      MethodHandle method, Object thisObject, Object[] arguments, int hookId) throws Throwable {
+    if (evalWithBindings == null) {
+      evalWithBindings =
+          methodFromObject(thisObject, "eval", "java.lang.String", "javax.script.Bindings");
+    }
+    if (evalWithBindings.isPresent()) {
+      String content = checkScriptContent(arguments[0], hookId);
+      return evalWithBindings.get().invoke(thisObject, content, arguments[1]);
+    } else {
+      return method.invokeWithArguments(thisObject, arguments);
+    }
+  }
+
+  private static String checkScriptContent(Object argument, int hookId) {
+    String script = null;
+    if (argument != null) {
+      if (argument instanceof String) {
+        script = (String) argument;
+      } else {
+        // This assumes that the reader content can safely be read completely
+        // and used as string parameter later on. Special reader
+        // implementations could be problematic, but are very unlikely to be
+        // used in this context.
+        script = readAll((Reader) argument);
+      }
+
+      if (script.contains(PAYLOAD)) {
+        Jazzer.reportFindingFromHook(new FuzzerSecurityIssueCritical(
+            "Script Engine Injection: Insecure user input was used in script engine invocation.\n"
+            + "Depending on the script engine's capabilities this could lead to sandbox escape and remote code execution."));
+      }
+      Jazzer.guideTowardsContainment(script, PAYLOAD, hookId);
+    }
+    return script;
+  }
+
+  private static String readAll(Reader reader) {
+    return new BufferedReader(reader).lines().collect(Collectors.joining());
```

Comment:
This drops line endings. We should probably directly read bytes from the reader instead, possibly limiting the buffet size to something reasonable.

Reply:
Right, changed.